### PR TITLE
fix: make GitHub Actions setup flow idempotent and stop bundling local changes

### LINKED ISSUE #904: Setup flow opened duplicate PRs and bundled unrelated local changes
## Summary

The React Doctor setup flow opened two separate PRs for the same requested GitHub Actions integration and bundled unrelated local application changes into those PR branches.

## Evidence

Target repo: https://github.com/hexclave/stackframe-website-2026

Two PRs were created for the same title within ~25 minutes:

- https://github.com/hexclave/stackframe-website-2026/pull/18
  - Branch: `react-doctor/add-github-actions`
  - Created: 2026-06-19T17:04:58Z
  - Commit: `84f6b36b4ceee0278dceb80121c94387b65e7737`
  - Added `.github/workflows/react-doctor.yml`, but also modified app/source files and deleted `pnpm-lock.yaml` entirely.
- https://github.com/hexclave/stackframe-website-2026/pull/19
  - Branch: `react-doctor/add-github-actions-202606191729`
  - Created: 2026-06-19T17:29:55Z
  - Commit: `5c4fd2755090e4497cf6fa9c0272da6aa978a045`
  - Same PR title and workflow file, plus additional local component/source changes and a large `pnpm-lock.yaml` rewrite.

The second PR appears to be a retry/superset of the first, but the first PR remained open and both branches contained much more than the React Doctor integration.

## Expected behavior

- Create or update one setup PR/branch for the React Doctor integration.
- Only include the files required for React Doctor setup, or clearly ask before bundling unrelated local worktree changes.
- Never delete or rewrite the package lockfile as an incidental setup artifact unless that is required and explained.

## Actual behavior

- Two duplicate PRs were opened.
- Local/unrelated application changes were swept into the generated PR branches.
- Cleanup required manually consolidating the intended changes into a local staged diff and closing both stray PRs.

### PR DESCRIPTION
**Pull request by rayhanadev**
## Root cause

Closes #904. The GitHub Actions setup flow opened two PRs for the same integration and swept unrelated local changes (modified app files, a deleted `pnpm-lock.yaml`) into the generated branches. Two independent bugs in `packages/react-doctor/src/cli/utils/open-workflow-pull-request.ts`:

### Bug 1 — duplicate PRs (idempotency)
The flow never checked for an already-open setup PR. `findUniqueBranchName` only avoided a *local* branch collision by appending a timestamp (`react-doctor/add-github-actions-202606191729` — the exact suffix format seen in the issue). Because the workflow file lives only on the prior PR's branch (not the user's working branch), a re-run re-writes it and opens a *second* PR.

### Bug 2 — bundled local changes
`git checkout -b <new> origin/<default>` carries the index onto the new branch, and the path-less `git commit` then commits the **whole index** — so anything the user had staged rode along into the PR. The inline comment assumed git refuses on conflict, but git only errors on *conflicts*; non-conflicting staged changes pass through silently.

## The fix

- **Idempotency:** before creating anything, `gh pr list` for an open PR whose head branch is under the `react-doctor/add-github-actions` prefix (covers timestamped variants too). If one exists, return the new `pr-exists` status and surface its URL instead of opening a duplicate.
- **Dirty-tree guard:** bail with `not-attempted` / `working-tree-dirty` when the tree has tracked changes **other than the workflow file**, falling back to staging the workflow file. The workflow path is excluded via a `:!` pathspec so the v1→v2 **upgrade** flow (which modifies the tracked workflow in place) still opens its PR.
- The existing-PR check runs before the dirty check (the open PR is the more useful signal). Both `setUpGitHubActions` and the upgrade flow in `handoff-to-agent` handle `pr-exists`.
- The command runner and `gh`-availability probe are now injectable for hermetic unit tests.

## Why this over #905

#905 (closed) added the dirty-tree guard only — a sound fix for Bug 2, but it leaves Bug 1 open: re-running on a *clean* tree still opens a second duplicate PR. This PR guards both, and scopes the dirty check so it doesn't break the upgrade flow.

## Testing

- New `tests/open-workflow-pull-request.test.ts` (11 cases): happy path, existing-PR short-circuit (canonical + timestamped + non-matching heads), dirty-tree bail, untracked-files allowed, workflow-path exclusion, ordering (PR check before dirty check), unparsable `gh` output, `branch-pushed`, and `gh-not-installed` short-circuit.
- `pnpm --filter react-doctor typecheck`, `pnpm lint`, `pnpm format:check` all green.

> The one flaky `install-react-doctor` persisted-decision test fails identically on `main` (it's `conf`-store state pollution from `vp` running each file under two vite configs) — unrelated to this change.

No new CLI flags, config, or JSON-report changes — only two new internal result variants and their messages.

---

> [!NOTE]
> **Medium Risk**
> Changes git/gh automation that creates branches and PRs; mistakes could skip setup or still pollute PRs, but behavior is guarded with fallbacks to staging and covered by new tests.
> 
> **Overview**
> Fixes **#904** by hardening `openWorkflowPullRequest` so re-running the GitHub Actions setup does not open a second PR and does not sweep unrelated local git changes into the automation branch.
> 
> **Idempotency:** Before any branch/checkout work, the CLI scans open PRs (`gh pr list`, capped via `GH_PR_LIST_MAX`) for a head branch under `react-doctor/add-github-actions` (including timestamped variants). If one exists, it returns **`pr-exists`** with that URL and skips all git writes. Fresh install (`setUpGitHubActions`) and the v1→v2 upgrade (`handoff-to-agent`) surface that message; the upgrade path also **reverts** the in-memory `@v2` edit so the working tree stays on `@v1` when the bump already lives on the open PR branch.
> 
> **Dirty-tree guard:** If tracked changes exist outside the workflow file (workflow path excluded via `:!` pathspec so in-place upgrades still work), the flow returns **`not-attempted` / `working-tree-dirty`** and callers fall back to **staging only the workflow file**. Untracked files are ignored; a failed status probe is treated as dirty. The existing-PR check runs before the dirty check.
> 
> **Other:** Injectable `run` / `gh` availability for tests; workflow-relative paths normalized with `toForwardSlashes` on Windows. New unit tests cover happy path, duplicate PR short-circuit, dirty tree, ordering, and edge cases.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit e7b80b31f11d066b69fa33098717cc8105ba2940. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

**Comment by rayhanadev** [comment]
## Adversarial review + deslop pass

Ran a multi-agent review (5 lenses: conventions, deslop, duplication, correctness, tests) with a skeptic adversarially verifying each finding. 6 survived and are now fixed in `34c164c4`:

- **Windows correctness bug** — `workflowRelative` came from `path.relative`, which yields backslashes on Windows; git's `:!` exclude pathspec doesn't treat those as separators, so the upgrade flow (workflow modified in place) would always hit `working-tree-dirty` and never open a PR on Windows. Now normalized via the repo's `toForwardSlashes`.
- **pr-exists end-state** — the upgrade flow's new `pr-exists` branch left the `@v2` edit uncommitted in the working tree; now restored to `@v1` to match the pr-opened path's "the bump lives only on the PR branch" contract.
- **Magic number** — extracted the `gh pr list` cap to `GH_PR_LIST_MAX` in `constants.ts` (AGENTS.md).
- **Test hardening** — the dirty-tree test now asserts add/commit/push were all skipped (not just checkout); the happy path now proves staging is path-scoped to the workflow file (no `git add -A`/`.`); branch-pushed now asserts the branch survives (`no git branch -D`) and HEAD is restored.

No new duplication surfaced (truffler search via the duplication lens came back clean).

File: packages/react-doctor/src/cli/utils/open-workflow-pull-request.ts [hunk 1]
```diff
@@ -39,20 +53,72 @@ export type NotAttemptedReason =
...
+    ["pr", "list", "--state", "open", "--json", "headRefName,url", "--limit", "100"],
+    cwd,
+  );
+  if (!prList.success) return null;
+  try {
+    const openPullRequests: OpenPullRequestSummary[] = JSON.parse(prList.stdout);
+    const setupPullRequest = openPullRequests.find((pullRequest) =>
+      (pullRequest.headRefName ?? "").startsWith(NEW_BRANCH_PREFIX),
+    );
+    return setupPullRequest?.url ?? null;
```
[the remark this answers is not in the document]
Comment (rayhanadev) [reply packages/react-doctor/src/cli/utils/open-workflow-pull-request.ts]: Fixed in e7b80b31. `findExistingSetupPullRequest` now returns the matched PR object and the caller short-circuits on its **existence** (not URL truthiness); the URL is used only for the message (empty when `gh` omits it). Added a regression test ("short-circuits on a matching setup PR even when gh omits the url"). Good catch — `gh --json url` always populates it, but the #904 guarantee shouldn't depend on that.

File: packages/react-doctor/src/cli/utils/open-workflow-pull-request.ts [hunk 2]
```diff
@@ -39,20 +53,72 @@ export type NotAttemptedReason =
...
+// none is open. Matches any head branch under the setup prefix so it catches
+// both the canonical branch and earlier timestamped variants. Best-effort:
+// a failed or unparsable `gh` call returns null so setup still proceeds.
+const findExistingSetupPullRequest = async (
+  cwd: string,
+  run: CommandRunner,
+): Promise<string | null> => {
+  const prList = await run(
+    "gh",
+    ["pr", "list", "--state", "open", "--json", "headRefName,url", "--limit", "100"],
```
[the remark this answers is not in the document]
Comment (rayhanadev) [reply packages/react-doctor/src/cli/utils/open-workflow-pull-request.ts]: Fixed in 34c164c4 — extracted to `GH_PR_LIST_MAX = 100` in `constants.ts` and referenced via `String(GH_PR_LIST_MAX)`. Thanks!

File: packages/react-doctor/src/cli/utils/handoff-to-agent.ts [hunk 3]
```diff
@@ -89,6 +89,10 @@ const upgradeGitHubActionsWorkflow = async (
     upgradeSpinner.succeed(
       `Opened pull request for review: ${highlighter.info(pullRequestResult.url)}`,
     );
+  } else if (pullRequestResult.status === "pr-exists") {
+    upgradeSpinner.succeed(
+      `A React Doctor pull request is already open: ${highlighter.info(pullRequestResult.url)}`,
+    );
```
[the remark this answers is not in the document]
Comment (rayhanadev) [reply packages/react-doctor/src/cli/utils/handoff-to-agent.ts:103]: Partly fixed, partly intentional.

**Leftover working-tree edit — fixed in 34c164c4.** The `pr-exists` branch now restores the original @v1 content (`fs.writeFileSync(workflow.workflowPath, workflow.content)`), so the file is no longer left silently modified/unstaged — matching the pr-opened path's "the bump lives only on the PR branch" contract.

**Recording the decision as `accepted` — intentional.** When the upgrade flow hits `pr-exists`, a React Doctor setup PR is already open under the `react-doctor/add-github-actions` prefix. Re-prompting on the next scan would just open another PR that immediately hits `pr-exists` again — a nag loop with no new outcome. Suppressing the offer is the right UX, and it's consistent with the pr-opened path: in both cases an open setup PR exists, so if the user later closes it without merging they stay on @v1 either way. The realistic `pr-exists` case in the upgrade flow *is* the upgrade PR from a prior run. Leaving this as-is.

[the remark this answers is not in the document]
Comment (rayhanadev) [reply packages/react-doctor/src/cli/utils/handoff-to-agent.ts:103]: Partly fixed, partly intentional.

**Leftover working-tree edit — fixed in 34c164c4.** The `pr-exists` branch now restores the original @v1 content (`fs.writeFileSync(workflow.workflowPath, workflow.content)`), so the file is no longer left silently modified/unstaged — matching the pr-opened path's "the bump lives only on the PR branch" contract.

**Recording the decision as `accepted` — intentional.** When the upgrade flow hits `pr-exists`, a React Doctor setup PR is already open under the `react-doctor/add-github-actions` prefix. Re-prompting on the next scan would just open another PR that immediately hits `pr-exists` again — a nag loop with no new outcome. Suppressing the offer is the right UX, and it's consistent with the pr-opened path: in both cases an open setup PR exists, so if the user later closes it without merging they stay on @v1 either way. The realistic `pr-exists` case in the upgrade flow *is* the upgrade PR from a prior run. Leaving this as-is.
